SU2_CFD/include/gradients/computeGradientsLeastSquares.hpp

```diff
@@ -0,0 +1,315 @@
+
+      for (size_t iVar = varBegin; iVar < varEnd; ++iVar)
+        for (size_t iDim = 0; iDim < nDim; ++iDim)
+          gradient(iPoint, iVar, iDim) = 0.0;
+
+      for (size_t iDim = 0; iDim < nDim; ++iDim)
+        for (size_t jDim = 0; jDim < nDim; ++jDim)
+          Rmatrix(iPoint, iDim, jDim) = 0.0;
+
+
+      for (size_t iNeigh = 0; iNeigh < node->GetnPoint(); ++iNeigh)
+      {
+        size_t jPoint = node->GetPoint(iNeigh);
+
+        const su2double* coord_j = geometry.node[jPoint]->GetCoord();
+        AD::SetPreaccIn(coord_j, nDim);
+
+        /*--- Distance vector from iPoint to jPoint ---*/
+
+        su2double dist_ij[MAXNDIM];
+
+        for (size_t iDim = 0; iDim < nDim; ++iDim)
+          dist_ij[iDim] = coord_j[iDim] - coord_i[iDim];
+
+        /*--- Compute inverse weight, default 1 (unweighted). ---*/
+
+        su2double weight = 1.0;
+
+        if (weighted)
+        {
+          weight = 0.0;
+          for (size_t iDim = 0; iDim < nDim; ++iDim)
+            weight += dist_ij[iDim] * dist_ij[iDim];
+        }
+
+        /*--- Sumations for entries of upper triangular matrix R. ---*/
+
+        if (weight > 0.0)
+        {
+          weight = 1.0 / weight;
+
+          Rmatrix(iPoint,0,0) += dist_ij[0]*dist_ij[0]*weight;
+          Rmatrix(iPoint,0,1) += dist_ij[0]*dist_ij[1]*weight;
+          Rmatrix(iPoint,1,1) += dist_ij[1]*dist_ij[1]*weight;
+
+          if (nDim == 3)
+          {
+            Rmatrix(iPoint,0,2) += dist_ij[0]*dist_ij[2]*weight;
+            Rmatrix(iPoint,1,2) += dist_ij[1]*dist_ij[2]*weight;
+            Rmatrix(iPoint,2,1) += dist_ij[0]*dist_ij[2]*weight;
+            Rmatrix(iPoint,2,2) += dist_ij[2]*dist_ij[2]*weight;
+          }
+
+          /*--- Entries of c:= transpose(A)*b ---*/
+
+          for (size_t iVar = varBegin; iVar < varEnd; ++iVar)
+          {
+            AD::SetPreaccIn(field(jPoint,iVar));
+
+            su2double delta_ij = weight * (field(jPoint,iVar) - field(iPoint,iVar));
+
+            for (size_t iDim = 0; iDim < nDim; ++iDim)
+              gradient(iPoint, iVar, iDim) += dist_ij[iDim] * delta_ij;
+          }
+        }
+      }
+
+      for (size_t iDim = 0; iDim < nDim; ++iDim)
+        for (size_t jDim = 0; jDim < nDim; ++jDim)
+          AD::SetPreaccOut(Rmatrix(iPoint, iDim, jDim));
+
+      for (size_t iVar = varBegin; iVar < varEnd; ++iVar)
+        for (size_t iDim = 0; iDim < nDim; ++iDim)
+          AD::SetPreaccOut(gradient(iPoint, iVar, iDim));
+
+      AD::EndPreacc();
+    }
+
+    /*--- Correct the gradient values across any periodic boundaries. ---*/
+
+    if (solver != nullptr)
+    {
+      SU2_OMP_MASTER
+      {
+        for (size_t iPeriodic = 1; iPeriodic <= config.GetnMarker_Periodic()/2; ++iPeriodic)
+        {
+          solver->InitiatePeriodicComms(&geometry, &config, iPeriodic, kindPeriodicComm);
+          solver->CompletePeriodicComms(&geometry, &config, iPeriodic, kindPeriodicComm);
+        }
+      }
+      SU2_OMP_BARRIER
```

Comment:
All threads must arrive here before execution continues, more important this gives all threads a coherent view of memory, which is fundamental here since only the master thread was involved in communicating halo values (without the barrier the other threads might not "see" the halo values correctly).
There are implicit barriers at the end of normal work sharing constructs, they can be removed by specifying the `nowait` clause, this is done in CSysVector methods (I think I explained why there, near the top of the cpp).

Reply:
thanks for the additional explanation of the OpenMP routines